template metaprogramming version of the drake c++ system class hierarchy

RussTedrake opened the pull request:
works for the pendulum example (runPendulumDynamics, runPendulumEnergyShaping, runPendulumLQR), including BotVisualizer output.  

There are places where we can clean up the syntax a bit, but I want to get people's thoughts on this now in prep for the design review/discussions happening in the next few days.

tkoolen commented on drake/core/Core.h:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    Vector1<ScalarType> first(void) const { return vec1; }
+    Vector2<ScalarType> second(void) const { return vec2; }
```
Should these return (const?) references?

RussTedrake replied:
definitely yes.

tkoolen commented on drake/core/Core.h:
```diff
@@ -0,0 +1,106 @@
+    Vector2<ScalarType> vec2;
+  };
+
+  template <template <typename> class Vector1, template <typename> class Vector2>
```
Although we never use the scalar type parameters for Vector1 and Vector2, it might still be a good idea to give them a name in such cases for readability.

RussTedrake replied:
i personally prefer this syntax if the names are not used (but it's not a strong opinion).

tkoolen commented on drake/examples/Pendulum/Pendulum.h:
```diff
@@ -3,100 +3,137 @@
-  }
-  virtual Drake::TaylorVecX dynamics(Drake::TaylorVarX t, const Drake::TaylorVecX& x, const Drake::TaylorVecX& u) const override {
-    return dynamics_implementation(t,x,u);
+  PendulumState& operator=(const Eigen::Matrix<ScalarType,2,1>& x) {
```
Should these assignment operators be templated on Derived and take a `MatrixBase<Derived>`? This wouldn't work for a `Block`, for example. Either that or use `Ref`?

RussTedrake replied:
fixed.

RussTedrake replied:
Do you have any suggestions for how to improve the typecast back out.  The best I've found in some quick looking is to define 
```
  template <int Rows>
  operator Eigen::Matrix<ScalarType,Rows,1> () const {
    Eigen::Matrix<ScalarType,2,1> x;
    x << theta, thetadot;
    return x;
  }
```
so it at least can convert into a larger matrix (e.g. both Vector2d and VectorXd).

tkoolen replied:
I don't think it's necessary to have cast operators to different Eigen types. Just return the smallest type you know will fit the state vector (in this case a Vector2d, in some cases it will need to be a VectorXd). If you need it to be a VectorXd, just construct a VectorXd using the Vector2d that comes out.

tkoolen replied:
So I think what you have now is fine.

tkoolen commented on drake/examples/Pendulum/Pendulum.h:
```diff
@@ -3,100 +3,137 @@
 
-class PendulumEnergyShaping : public DrakeSystem {
+
+class PendulumEnergyShaping : public Drake::System<PendulumEnergyShaping,Drake::UnusedVector,PendulumState,PendulumInput,false,true> {
```
I know it's been called PendulumEnergyShaping on the Matlab side forever, but I still think that PendulumEnergyShapingController would be clearer.

RussTedrake replied:
ok with me.

tkoolen commented on drake/examples/Pendulum/runEnergyShaping.cpp:
```diff
@@ -1,17 +1,43 @@
+    Drake::CombinedVectorBuilder<PendulumState,PendulumInput>::VecType<double> test(abc);
+    cout << test << endl;
+    test=2*abc;
+    cout << test << endl;
```
Should actually assert something instead of printing.

RussTedrake replied:
done.

tkoolen commented on drake/solvers/DrakeFunction.h:
```diff
@@ -0,0 +1,42 @@
+  template <typename Derived, template<typename> class InputVector, template<typename> class OutputVector >
+  class Function {
+  public:
+    virtual OutputVector<double> operator()(const InputVector<double>& x)  { // todo: add const (w/o compile error)?
```
I still think this shouldn't be virtual.

RussTedrake replied:
these were old drafts...

tkoolen commented on drake/solvers/test/simpleFunctionTest.cpp:
```diff
@@ -0,0 +1,167 @@
+using namespace std;
+
+template <typename ScalarType = double>
+class PendulumState  { // models the Drake::Vector concept
```
Duplicate PendulumState?

tkoolen replied:
In fact, should this whole file be removed?

tkoolen commented on drake/systems/DrakeSystem.h:
```diff
@@ -136,7 +126,7 @@ class DLLEXPORT FeedbackSystem : public DrakeSystem {
 
   virtual bool isTimeInvariant() const override { return sys1->isTimeInvariant() && sys2->isTimeInvariant(); }
-  virtual bool isDirectFeedthrough() const override { return sys1->isDirectFeedthrough() && sys2->isDirectFeedthrough(); }
+  virtual bool isDirectFeedthrough() const override { return sys1->isDirectFeedthrough(); } // && sys2->isDirectFeedthrough(); }
```
Why?

tkoolen commented on drake/systems/System.h:
```diff
@@ -0,0 +1,242 @@
+    const static bool isDirectFeedthrough = _isDirectFeedthrough;
+    const static unsigned int num_states = VectorTraits<StateVector<double> >::RowsAtCompileTime;
+    const static unsigned int num_inputs = VectorTraits<InputVector<double> >::RowsAtCompileTime;
+    const static unsigned int num_outputs = VectorTraits<OutputVector<double> >::RowsAtCompileTime;
```
I think we want to use int instead of unsigned int. Certainly Eigen uses the same enum trick for e.g. RowsAtCompileTime, and they can have a negative number of rows at compile (-1 to indicate dynamic size). Note that the actual underlying type of an enum is not specified by the standard (https://stackoverflow.com/questions/159034/are-c-enums-signed-or-unsigned). The best way to get around it I think is to use constexpr instead of the enum trick in VectorTraits, but I believe that MSVC 2013 doesn't fully support constexpr yet.

tkoolen replied:
Also, having these unsigned means that we can't use a special value to indicate that the number of states is not known at compile time. I think requiring that the number of states is always known at compile time is an unreasonable requirement.

RussTedrake replied:
but what about systems that have 2^31 < num_states < 2^32  ?  :)  
i've made the change.

tkoolen commented on drake/systems/SystemSpecializations.h:
```diff
@@ -0,0 +1,75 @@
+  OutputDispatchSpecialization(false, false, true, sys->outputImplementation(u))
+  OutputDispatchSpecialization(false, false, false, sys->outputImplementation())
+
+};
```
`#undef OutputDispatchSpecialization`?

RussTedrake replied:
good.

tkoolen commented on drake/systems/SystemSpecializations.h:
```diff
@@ -0,0 +1,75 @@
+    }
+  };
+
+#define OutputDispatchSpecialization(time, state, input, func_call) \
```
Perhaps prepend Drake_? Should make sure we don't clobber someone else's macro.

tkoolen commented on drake/systems/plants/BotVisualizer.h:
```diff
@@ -11,125 +11,133 @@
+
+      Eigen::Matrix<double,VectorTraits<InputVector<double> >::RowsAtCompileTime,1> uvec(u);
+      auto q = uvec.head(manip.num_positions);
+      KinematicsCache<double> cache = manip.doKinematics(q);
```
Would be a lot more efficient to create a KinematicsCache as a member variable and reuse it.

tkoolen replied:
Ah but then that would have to be mutable too I guess. It's unfortunate that we have to cheat a little like this to make things efficient.

tkoolen commented on drake/systems/plants/BotVisualizer.h:
```diff
@@ -11,125 +11,133 @@
-    return Eigen::VectorXd::Zero(0);
-  }
+  private:
+    mutable RigidBodyManipulator manip;  // todo: remove mutable tag after RBM cleanup
```
Is this still necessary?

RussTedrake replied:
yes.  it seems so.

tkoolen commented on drake/systems/System.h:
```diff
@@ -0,0 +1,242 @@
+//          CombinedVectorBuilder<System1::StateVectorType , System2::StateVectorType>::VecType, System1::InputVectorType, System1::OutputVectorType, System1::isTimeVarying||System2::isTimeVarying, System1::isDirectFeedthrough> {
+//
+// So I'm getting around it with this helper macro (which makes the call outside of the template class, so it's not recursive)
+#define FeedbackSystemType(System1,System2) Drake::FeedbackSystem<System1,System2,System1::StateVectorType,System2::StateVectorType,Drake::CombinedVectorBuilder<System1::StateVectorType,System2::StateVectorType>::VecType,System1::InputVectorType,System1::OutputVectorType,System1::isTimeVarying||System2::isTimeVarying,System1::isDirectFeedthrough>
```
I'm going to take a crack at doing this without a macro. A major issue with this macro is currently that if System1 has multiple type parameters, it will look like `Foo<Bar, Baz>`. The comma will be interpreted in the macro expansion as the argument separator, so it will think that `, Baz>` is System2 and that you're passing in more arguments than it expects.

tkoolen commented on drake/core/Core.h:
```diff
@@ -0,0 +1,106 @@
+
+
+  template <typename ScalarType, template <typename> class Vector1, template <typename> class Vector2>
+  class CombinedVector {
```
I wonder if this should be variadically templated so that it can handle concatenations of arbitrarily many types.

In fact, isn't CombinedVector just a std::tuple with an assignment operator that takes an eigen matrix?

RussTedrake replied:
happy to revisit in the future.  but will not make this change yet.

RussTedrake commented:
hopefully this is good to go now (will check the builds again first thing in the morning).
i think it's pretty clean, although I have a few ideas to make it even cleaner.

the biggest thing that is missing is that I'm not done re-implementing the c++ runLCM functionality.  I do have a version that's pretty close.  @psiorx -- is it ok to merge this if the quadrotor is not publishing LCM messages (except to the visualizer)?

@tkoolen -- if you have the energy for another glance through the code, that would be great.

tkoolen commented on drake/CMakeLists.txt:
```diff
@@ -108,6 +108,15 @@ mex_setup()
 #   message("Setting CTEST Configuration Type to ${CTEST_CONFIGURATION_TYPE}")
 #endif()
 
+file(WRITE ${PROJECT_SOURCE_DIR}/core/Path.h
```
Not necessary for now, but would it be an idea to put this in the PROJECT_BINARY_DIR instead of PROJECT_SOURCE_DIR, since it's a build product? That's what CMake's GenerateExportHeader does by default as well.

tkoolen commented on drake/examples/Pendulum/CMakeLists.txt:
```diff
@@ -8,27 +9,33 @@ include_directories(${PROJECT_SOURCE_DIR}/util)
   add_executable(runPendulumLQR runLQR.cpp)
-  target_link_libraries(runPendulumLQR drakeSystem drakeRBM)
+  target_link_libraries(runPendulumLQR drakeRBM)
+  add_dependencies(runPendulumLQR drakeUtil drake_lcmtypes lcmtype_agg_hpp)
```
Won't drakeUtil already be a dependency due to linking against drakeRBM?

tkoolen commented on drake/examples/Pendulum/CMakeLists.txt:
```diff
@@ -8,27 +9,33 @@ include_directories(${PROJECT_SOURCE_DIR}/util)
-  target_link_libraries(runPendulumDynamics drakeSystem drakeRBM)
+  target_link_libraries(runPendulumDynamics drakeRBM)
+  add_dependencies(runPendulumDynamics drake_lcmtypes lcmtype_agg_hpp)
+  pods_use_pkg_config_packages(runPendulumDynamics lcm)
```
We should have a macro for the above two lines. I know haven't been doing this right myself.

tkoolen commented on drake/examples/Pendulum/Pendulum.h:
```diff
@@ -3,100 +3,128 @@
+  enum {
+    RowsAtCompileTime = 2
+  };
+  static std::size_t size() { return 2; }
```
I don't think we can always have this be static. For an RBM loaded from a URDF, we won't know the size until runtime and we need to pick a state vector type at compile time, so the size would probably be passed into the constructor of the state vector type, meaning it isn't necessarily the same for all instances of the class.

We can fix that later though.

tkoolen commented on drake/examples/Pendulum/Pendulum.h:
```diff
@@ -3,100 +3,128 @@
+  enum {
+    RowsAtCompileTime = 1
+  };
+  static unsigned int size() { return 1; }
```
size_t here too for consistency?

tkoolen commented on drake/systems/LQR.h:
```diff
@@ -0,0 +1,52 @@
+  template <typename Derived, template<typename> class StateVector, template<typename> class InputVector, template<typename> class OutputVector, bool isTimeVarying, bool isDirectFeedthrough>
+  std::shared_ptr<AffineSystem<UnusedVector,StateVector,InputVector> > timeInvariantLQR(const System<Derived,StateVector,InputVector,OutputVector,isTimeVarying,isDirectFeedthrough>& sys, const StateVector<double>& x0, const InputVector<double>& u0, const Eigen::MatrixXd& Q, const Eigen::MatrixXd& R) {
+    const unsigned int num_states = StateVector<double>::RowsAtCompileTime;
+    const unsigned int num_inputs = InputVector<double>::RowsAtCompileTime;
```
Shouldn't be unsigned. Also need to handle case where `StateVector<double>::RowsAtCompileTime != x0.size()`, same for inputs.

tkoolen commented on drake/systems/LinearSystem.h:
```diff
@@ -1,38 +1,64 @@
+  template <typename ScalarType>
+  StateVector<ScalarType> dynamicsImplementation(const StateVector<ScalarType>& x, const InputVector<ScalarType>& u) const {
+    using namespace Eigen;
+    StateVector<ScalarType> xdot = A*static_cast<Matrix<ScalarType,num_states,1> >(x) + B*static_cast<Matrix<ScalarType,num_inputs,1> >(u) + xdot0;
```
This kind of usage makes me think that it might be better to just require that for any state vector there exists a toMatrix function that returns an eigen matrix of the appropriate size.

For later.

tkoolen commented on drake/systems/System.h:
```diff
@@ -0,0 +1,265 @@
+    const static bool isDirectFeedthrough = _isDirectFeedthrough;
+    const static int num_states = StateVector<double>::RowsAtCompileTime;
+    const static int num_inputs = InputVector<double>::RowsAtCompileTime;
+    const static int num_outputs = OutputVector<double>::RowsAtCompileTime;
```
Add _at_compile_time to avoid confusion?

tkoolen commented on drake/systems/System.h:
```diff
@@ -0,0 +1,265 @@
+  };
+
+
+#define CascadeSystemType(System1,System2) Drake::CascadeSystem<System1,System2,System1::StateVectorType,System2::StateVectorType,,>
```
Get rid of this now?

tkoolen commented:
Done with my review.

psiorx commented:
I'm OK with pointing to my own branch until the LCM code is finalized.  I
would be happy to convert the quad example to the new framework.
On Dec 13, 2015 10:13 PM, "Russ Tedrake" <notifications@github.com> wrote:

> hopefully this is good to go now (will check the builds again first thing
> in the morning).
> i think it's pretty clean, although I have a few ideas to make it even
> cleaner.
>
> the biggest thing that is missing is that I'm not done re-implementing the
> c++ runLCM functionality. I do have a version that's pretty close. @psiorx
> <https://github.com/psiorx> -- is it ok to merge this if the quadrotor is
> not publishing LCM messages (except to the visualizer)?
>
> @tkoolen <https://github.com/tkoolen> -- if you have the energy for
> another glance through the code, that would be great.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/1512#issuecomment-164331333>
> .
>

RussTedrake commented:
awesome.  i've addressed all the extra comments.  thanks @tkoolen !
@psiorx -- i've already updated the quadrotor example.  but the ode tools are going to take some work.  :)

psiorx commented:
Just got a chance to take a closer look at this.   Looks great! 

>  i've already updated the quadrotor example

I see you've also done the C++ care solver as well.  Thanks.  

> but the ode tools are going to take some work

It looks like `odeint` relies on some of the rest of boost (config.hpp, etc) but it looks like the combined internal dependencies stay in the realm of header-only so no linking against boost libraries is necessary.  

I know we've tried to get away from boost in the past (specifically mex files that are linked against boost).  Are we willing to make an exception here since this fits so well with our new template metaprogramming approach?

RussTedrake commented:
@psiorx -- I've been avoiding boost still, despite it's obvious synergy with the metaprogramming.  I think that's a serious discussion that we need to have.
